### PR TITLE
Ensure code is in fragment for Angular guard

### LINKED ISSUE #4084: angular MsalGuard wrongly checking for 'code='  in URL
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.16.1

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.0.2

### Description

When the app url contains a query parameter named like "table_code=XXx", it creates a redirect loop due to MsalGuard.

It's probably because of the way checking for the string `code=` in the current path, the check should be stricter (like expecting the / or # or ? character before)

image

### Msal Logs

[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
[Wed, 22 Sep 2021 07:37:59 GMT] : @azure/msal-angular@2.0.2 : Info - Guard - Hash contains known code response, stopping navigation.

### MSAL Configuration

```javascript
new PublicClientApplication({
    auth: {
      clientId: xClientId,
      authority: 'https://login.microsoftonline.com/' + xTenandId,
      redirectUri: URL_BASE + '/login-redirect',
      postLogoutRedirectUri: '/'
    },
    cache: {
      cacheLocation: BrowserCacheLocation.LocalStorage,
      storeAuthStateInCookie: isIE, // set to true for IE 11. Remove this line to use Angular Universal
    },
    system: {
      loggerOptions: {
        loggerCallback,
        logLevel: LogLevel.Info,
        piiLoggingEnabled: false
      }
    }
  })
```

### Relevant Code Snippets

```javascript
export function MSALGuardConfigFactory(): MsalGuardConfiguration {
  return {
    interactionType: InteractionType.Redirect,
    authRequest: {
      scopes: [`api://${xClientId}/.default`]
    },
    loginFailedRoute: '/#login-failed'
  };
}

...
providers: [
    {
      provide: MSAL_GUARD_CONFIG,
      useFactory: MSALGuardConfigFactory
    },
```
```


### Reproduction Steps

1. Have a secured route query param named `table_code`
2. Navigate to route /secured-route?table_code=123
3. Browser page freezes due to redirect loop

### Expected Behavior

It should not enter a redirect loop when a query parameter name ends by `code`

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome, Edge, Other

### Regression

_No response_

### Source

External (Customer)

### PR DESCRIPTION
**Pull request by jasonnutter**
The MSAL Guard is currently incorrectly detecting `code=` when inside unrelated query parameters, instead of just looking in the fragment. This PR ensures the guard only looks for `code=` in the url fragment for path routing, and only looks in the url if hash routing it used.

### CODE REVIEW

File: lib/msal-angular/src/msal.guard.ts [hunk 1]
```diff
@@ -175,6 +193,12 @@ export class MsalGuard implements CanActivate, CanActivateChild, CanLoad {
             );
     }
 
+    includesCode(path: string): boolean {
+        return path.indexOf("/code") > -1 ||
```
Comment (tnorling) [inline lib/msal-angular/src/msal.guard.ts]: Does this need the =?
```suggestion
        return path.indexOf("/code=") > -1 ||
```

File: lib/msal-angular/src/msal.guard.ts [hunk 2]
```diff
@@ -144,16 +144,34 @@ export class MsalGuard implements CanActivate, CanActivateChild, CanLoad {
...
+                         */
+
+                        /*
+                         * Hash routing:
+                         * state.url: /code
+                         * state.root.fragment: null
+                         */
+                        const urlContainsCode = this.includesCode(state.url);
+                        const fragmentContainsCode = state.root && state.root.fragment && this.includesCode(`#${state.root.fragment}`);
+                        const hashRouting = this.location.prepareExternalUrl(state.url).startsWith("#");
```
Comment (tnorling) [inline lib/msal-angular/src/msal.guard.ts]: Is `.startsWith` a problem for IE11 compat?

Comment (jasonnutter) [reply lib/msal-angular/src/msal.guard.ts]: Oops, yes, thanks.

File: lib/msal-angular/src/msal.guard.ts (hunk 1 above)
Comment (jasonnutter) [reply lib/msal-angular/src/msal.guard.ts]: With hash routing, the parsed path does not include the `=` (see the comment in the guard above).

Comment (tnorling) [reply lib/msal-angular/src/msal.guard.ts]: How does that work? How are the keys and values separated?

Comment (pkanher617) [reply lib/msal-angular/src/msal.guard.ts]: I'm a bit confused about this as well.

Comment (jasonnutter) [reply lib/msal-angular/src/msal.guard.ts]: > How does that work? How are the keys and values separated?

Sorry, referring to how the Angular router parses the url, not MSAL. 

When the Angular router (with hash routing) sees `/#code=123`, it parses it to `/code`

Comment (jasonnutter) [reply lib/msal-angular/src/msal.guard.ts]: I guess this is assuming the app has a `/code` route registered. Lemme see what happens if a `/code` path isn't registered.

Comment (tnorling) [reply lib/msal-angular/src/msal.guard.ts]: Would that mean that someone who has a `codes` route or anything else that starts with `code` would have to use path routing to be compatible? If so, that should be documented.

Comment (jasonnutter) [reply lib/msal-angular/src/msal.guard.ts]: Hmm good point, will make this smarter such that `/codes` and `/code` are treated different.

Comment (jasonnutter) [reply lib/msal-angular/src/msal.guard.ts]: When using hash routing, `/code` must be registered, and updated logic to ensure route must end with `/code`.

**Review by tnorling — APPROVED** [review]

**Review by jo-arroyo — APPROVED** [review]

**Pull request merged**